src/models/semantic.py

```diff
@@ -817,22 +836,30 @@ def on_test_epoch_end(self) -> None:
             self.test_cm.reset()
             return
 
+        if self.trainer.num_devices > 1:
+            epoch_cm = torch.sum(self.all_gather(self.test_cm.confmat), dim=0)
+        else:
+            epoch_cm = self.test_cm
+        # Reset metrics accumulated over the last epoch
+        self.test_cm.reset()
+
+        epoch_cm = ConfusionMatrix(self.num_classes).from_confusion_matrix(epoch_cm)
+
         # Log metrics
-        self.log("test/miou", self.test_cm.miou(), prog_bar=True)
-        self.log("test/oa", self.test_cm.oa(), prog_bar=True)
-        self.log("test/macc", self.test_cm.macc(), prog_bar=True)
-        for iou, seen, name in zip(*self.test_cm.iou(), self.class_names):
+        self.log("test/miou", epoch_cm.miou(), prog_bar=True, rank_zero_only=True)
+        self.log("test/oa", epoch_cm.oa(), prog_bar=True, rank_zero_only=True)
+        self.log("test/macc", epoch_cm.macc(), prog_bar=True, rank_zero_only=True)
+        for iou, seen, name in zip(*epoch_cm.iou(), self.class_names):
             if seen:
-                self.log(f"test/iou_{name}", iou, prog_bar=True)
+                self.log(f"test/iou_{name}", iou, prog_bar=True, rank_zero_only=True)
 
         # Log confusion matrix to wandb
         if isinstance(self.logger, WandbLogger):
             self.logger.experiment.log({
                 "test/cm": wandb_confusion_matrix(
-                    self.test_cm.confmat, class_names=self.class_names)})
+                    epoch_cm.confmat, class_names=self.class_names)})
 
-        # Reset metrics accumulated over the last epoch
-        self.test_cm.reset()
```

Comment:
Why remove the call to .reset() ? Have you tested this still preserved the single-device behavior and metrics ?